patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs

```diff
@@ -0,0 +1,184 @@
+
+namespace Terraria.ModLoader.Recipes
+{
+	public class RecipeBuilder
+	{
+		private readonly ModRecipe _recipe;
+
+
+		/// <summary>Creates a new instance with no ingredients and no result.</summary>
+		/// <seealso cref="Requires(int,int)"/>
+		/// <seealso cref="Build"/>
+		public RecipeBuilder()
+		{
+		}
+
+		/// <summary>Creates a new instance with no ingredients with the given item and stack as a result.</summary>
+		/// <param name="mod">The mod who owns the recipe.</param>
+		/// <param name="item">The item type.</param>
+		/// <param name="stack">The stack.</param>
+		/// <seealso cref="Requires(int,int)"/>
+		/// <seealso cref="Build"/>
+		public RecipeBuilder(Mod mod, int item, int stack = 1)
+		{
+			_recipe = new ModRecipe(mod);
+			_recipe.SetResult(item, stack);
+		}
+
+		/// <summary>Creates a new instance with no ingredients with the given item and stack as a result.</summary>
+		/// <param name="mod">The mod who owns the recipe.</param>
+		/// <param name="itemName">Name of the item.</param>
+		/// <param name="stack">The stack.</param>
+		/// <seealso cref="Requires(int,int)"/>
+		/// <seealso cref="Build"/>
+		public RecipeBuilder(Mod mod, string itemName, int stack = 1)
+		{
+			_recipe = new ModRecipe(mod);
+			_recipe.SetResult(mod, itemName, stack);
+		}
+
+		/// <summary>Creates a new instance with no ingredients with the given item and stack as a result.</summary>
+		/// <param name="item"></param>
+		/// <param name="stack"></param>
+		/// <seealso cref="Requires(int,int)"/>
+		/// <seealso cref="Build"/>
+		public RecipeBuilder(ModItem item, int stack = 1)
+		{
+			_recipe = new ModRecipe(item.mod);
+			_recipe.SetResult(item, stack);
+		}
+
+
+		/// <summary>
+		/// Adds an ingredient to this recipe with the given item type and stack size.
+		/// Ex.: 
+		/// <example>recipe.AddIngredient(ItemID.IronAxe)</example>
+		/// </summary>
+		/// <param name="type">The item type.</param>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		public RecipeBuilder Requires(int type, int stack = 1)
+		{
+			_recipe.AddIngredient(type, stack);
+
+			return this;
+		}
+
+
+		// System.TupleValue is not resolved. Need to import it from a NuGet package ?
+		/*public RecipeBuilder Requires(params (short itemId, int stack)[] ingredients)
+		{
+			for (int i = 0; i < ingredients.Length; i++)
+				Requires(ingredients[i].Item1, ingredients[i].Item2);
+
+			return this;
+		}*/
+
+
+		/// <summary>
+		/// Adds the specified ingredients to this recipe with the given item types.
+		/// Ex.: 
+		/// <example>recipe.AddIngredient(ItemID.IronAxe)</example>
+		/// </summary>
+		/// <param name="type1">The first item type.</param>
+		/// <param name="type2">The second item type.</param>
+		/// <param name="type3">The third item type.</param>
+		/// <param name="types">The remaining item types.</param>
+		/// <returns></returns>
+		public RecipeBuilder Requires(int type1, int type2, int type3, params int[] types)
```

Comment:
Why type1 through 3 and then params types? Just do params types.

Reply:
I did as was discussed on this [comment](https://github.com/tModLoader/tModLoader/issues/847#issuecomment-621730542), but I will change it back to `Requires(params int[])`.

Reply:
You put this back, by accident?

Reply:
Yea, I'll remove it again. Seems I forgot to fetch before pulling.